src/engraving/libmscore/textedit.h

```diff
@@ -60,13 +60,74 @@ struct TextEditData : public ElementEditData {
 class TextEditUndoCommand : public UndoCommand
 {
 protected:
-    TextCursor c;
+    TextCursor curs;
 public:
     TextEditUndoCommand(const TextCursor& tc)
-        : c(tc) {}
+        : curs(tc) {}
     bool isFiltered(UndoCommand::Filter f, const Element* target) const override
     {
-        return f == UndoCommand::Filter::TextEdit && c.text() == target;
+        return f == UndoCommand::Filter::TextEdit && curs.text() == target;
+    }
+
+    TextCursor& cursor() { return curs; }
+};
+
+//---------------------------------------------------------
+//   ChangeText
+//---------------------------------------------------------
+
+class ChangeTextProperties : public TextEditUndoCommand
+{
+    QString xmlText;
+    Pid propertyId;
+    QVariant propertyVal;
+    FontStyle existingStyle;
+
+    void restoreSelection()
+    {
+        TextCursor& tc = cursor();
+        tc.text()->cursor()->setSelectLine(tc.selectLine());
+        tc.text()->cursor()->setSelectColumn(tc.selectColumn());
+        tc.text()->cursor()->setRow(tc.row());
+        tc.text()->cursor()->setColumn(tc.column());
+    }
+
+public:
+    ChangeTextProperties(const TextCursor* tc, Ms::Pid propId, const QVariant& propVal)
+        : TextEditUndoCommand(*tc)
+    {
+        propertyId = propId;
+        propertyVal = propVal;
+        if (propertyId == Pid::FONT_STYLE) {
+            existingStyle = static_cast<FontStyle>(cursor().text()->getProperty(propId).toInt());
+        }
+    }
+
+    virtual void undo(EditData*) override
+    {
+        cursor().text()->setXmlText(xmlText);
+        restoreSelection();
+        cursor().text()->layout();
+    }
+
+    virtual void redo(EditData*) override
+    {
+        xmlText = cursor().text()->xmlText();
+        restoreSelection();
+        if (propertyId == Pid::FONT_STYLE) {
+            FontStyle setStyle = static_cast<FontStyle>(propertyVal.toInt());
+            if ((setStyle& FontStyle::Bold) != (existingStyle & FontStyle::Bold)) {
+                cursor().setFormat(FormatId::Bold, setStyle & FontStyle::Bold);
```

Comment:
all this logic is because bold/italic/underline aren't properties in their own right, but we need to know which one is being set, to ensure we only apply that and don't override the others